### DISCUSSION TITLE
Noticed some rather user-unfriendly behavior regarding remote backend and parameters saving.

### DISCUSSION THREAD

**Discussion by sguart** [mcmonkeyprojects/SwarmUI#284] [Q&A]
Below is an issue that I encountered, since I don't know exactly the implementation, so all of the below is just my speculation and it's up for discussion.
 
It seems like some kind of information related to the backend ID is saved along with the generated image.

So, if the backend configuration is altered, and the user clicked on "Resue Parameters" then generate an image. The image is generated and you can click on it the web page, but it's not saved to the output folder. The log shows no error and it appeared to find the right backend to generate an image but output is no longer save.

To recover from this, the user will have to "Reuse Parameters" from a previously generated image that was created without the backend modification to be able to generate the image and have the output saved like the normal flow.

One quick way to recover from this is to use the Simple tab to generate an image and "Reuse Parameters" from that image then change the parameters to resume generating the image. This acts as a setting reset to clear out the backend related parameters.

Restart the server or perform a clean install doesn't resolve this if the user kept previously generated image with the invalid backend configuration and use "Reuse Parameters" to generate image.

There's no obvious error in the log and it still generate an image but it's just not saving it any more. Not sure if this is comfy or swarm domain.

This issue of backend ID related info saved with the generated image also manifest into another problem even with the backend configuration intact.

Let's say you were creating a remote backend but you didn't know the exact URL so you create one with the wrong one and then you noticed it and edit with the correct URL. In the background, it will create a backend with a different ID. 

Then let's say it worked now and you generated image using one local and one remote and it's a success. Now, you restart your server, the backend ID will get rearrange. So, if you "Reuse Parameter" on that last image and try to generate an image. It will no longer use your remote backend. It will silently failed. When you isolate the backend to force it to generate image thru it, it will then tell you that it failed because it can't find the backend ID.

To recover from this, you will need do step similiar to the above, ie generate an image thru the Simple tab and then "Reuse Parameter" from that image to clear out the setting related to the backend ID, or use a previous image that was generated without the additonal backend configured.

Maybe these situations can be avoided, if the server have a clear all settings button, as it seems to be some settings that's not visible to the user to modify, and seems to be related to backend config.
 
Welcome more discussion for people that have extensive experience with remote backends.

thank you so much

**Comment by sguart** [comment]
didn't notice there is a parameter called Exact BackendID that's set and can be altered.

**Reply by sguart** [reply]
actually unsetting this parameter and generate image result in the same behavior of image not being saved.

**Reply by sguart** [reply]
okay more testing revealed that the symptom only happens when the generated image is saved with Exact BackendID. If "Reuse Parameters" is used on this image, it resulted in no image being save once your backend configuration is no longer the same as when the image was generated. And, unsetting that parameter from the settings and generate image doesn't improve the situation.

**Comment by mcmonkey4eva** [comment]
at the top right there is Quick Tools which has a Reset All Parameters. You can also select to view only modified parameters to see what it is you've altered.

Exact Backend ID does not set itself, you set that.

You may have also set DoNotSave from the sounds of it? Images always save unless you disable saving. If you have an invalid custom backend ID it'll just not generate at all and return an error.

**Reply by sguart** [reply]
Yeah I see that now, Quck Tools -> Reset Parameter to Default. Thanks for the tip.

Regarding the other two points that you brought up, unfortunately, I was new to the tool, so, I wasn't even aware of the "DoNotSave" and "Exact Backend ID" parameters. So, I never even attempted to use them at all.

So, in my experience, "Exact Backend ID" does set itself as I never attempt to set it since I don't know its existence. "DoNotSave" is never set also, since I don't know about it. But it seems like "DoNotSave" is not a factor here.

***More Testing***

Ah, so I just tested it and "Exact Backend ID:0" and "DoNotSave:true" it automatically added to the image that was sent to the remote backend.

Actually that's not exactly true. This auto add "Exact Backend ID:0" and "DoNotSave:true" actually happened after I disabled the remote backend then generate a single image.

Previously I generated 3 images. Then I disabled remote backend. Then generate 1 image.

Now, two of the 3 previously generated images have "Exact Backend ID:0" and "DoNotSave:true" on them. So, it retroactively updated previously generated images save parameters.

Now, I re-enable the remote backend, and generate 3 images again. The 2nd image that was sent to the remote backend automatically have the "Exact Backend ID:0" and "DoNotSave:true" added to it.

***More Testing***

Since I don't know about those parameters, and selected the remote image to "Reuse Parameters", which resulted in the behavior that I observed.

**Reply by mcmonkey4eva** [reply]
OH, I see - using a remote Swarm-API-Backend, it's mistakenly adding internal parameters for the backend routing that end up visible to you in the UI. This is a bug and should be filed as a bug on the issues page

**Reply by sguart** [reply]
Okay will file a bug repot. Thanks! :)
